python/pyspark/mllib/linalg/distributed.py

```diff
@@ -0,0 +1,513 @@
+            rows = rows_df.map(lambda row: IndexedRow(row[0], row[1]))
+            return IndexedRowMatrix(rows, java_matrix=java_matrix)
+        elif class_name == "CoordinateMatrix":
+            # We use DataFrames for serialization of MatrixEntry entries
+            # from Java, so we first convert the RDD of entries to a
+            # DataFrame on the Scala/Java side. Then we map each Row in
+            # the DataFrame back to a MatrixEntry on this side.
+            entries_df = callMLlibFunc("getMatrixEntries", java_matrix)
+            entries = entries_df.map(lambda row: MatrixEntry(row[0], row[1], row[2]))
+            return CoordinateMatrix(entries, java_matrix=java_matrix)
+        else:
+            raise TypeError("Cannot create distributed matrix from Java %s" % class_name)
+    else:
+        raise TypeError("java_matrix should be JavaObject, got %s" % type(java_matrix))
+
+
+class DistributedMatrix(object):
+    """
+    Represents a distributively stored matrix backed by one or
+    more RDDs.
+
+    """
+    def numRows(self):
+        """Get or compute the number of rows."""
+        raise NotImplementedError
+
+    def numCols(self):
+        """Get or compute the number of cols."""
+        raise NotImplementedError
+
+
+class RowMatrix(DistributedMatrix):
+    """
+    .. note:: Experimental
+
+    Represents a row-oriented distributed Matrix with no meaningful
+    row indices.
+
+    :param rows: An RDD of vectors.
+    :param numRows: Number of rows in the matrix. A non-positive
+                    value means unknown, at which point the number
+                    of rows will be determined by the number of
+                    records in the `rows` RDD.
+    :param numCols: Number of columns in the matrix. A non-positive
+                    value means unknown, at which point the number
+                    of columns will be determined by the size of
+                    the first row.
+    """
+    def __init__(self, rows, numRows=0, numCols=0, java_matrix=None):
+        """Create a wrapper over a Java RowMatrix."""
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD of vectors, got %s" % type(rows))
+        rows = rows.map(_convert_to_vector)
+
+        if not (isinstance(java_matrix, JavaObject)
+                and java_matrix.getClass().getSimpleName() == "RowMatrix"):
+            java_matrix = callMLlibFunc("createRowMatrix", rows, long(numRows), int(numCols))
+
+        self._java_matrix_wrapper = JavaModelWrapper(java_matrix)
+        self._rows = rows
+
+    @property
+    def rows(self):
+        """Rows of the RowMatrix stored as an RDD of vectors."""
+        return self._rows
+
+    def numRows(self):
+        """
+        Get or compute the number of rows.
+
+        >>> rows = sc.parallelize([[1, 2, 3], [4, 5, 6],
+        ...                        [7, 8, 9], [10, 11, 12]])
+
+        >>> rm = RowMatrix(rows)
+        >>> print(rm.numRows())
```

Comment:
So the issue here is that `numRows()` returns a long, so Python 2 outputs `4L` while Python 3 outputs `4`. We could wrap it in `int()` but I thought that looked more confusing for an end user looking at these tests in the API docs, so I selected `print()`.